File: src/main/java/it/robfrank/linklift/application/domain/model/GraphData.java
package it.robfrank.linklift.application.domain.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.List;
import org.jspecify.annotations.NonNull;

public record GraphData(@JsonProperty("nodes") @NonNull List<LinkNode> nodes, @JsonProperty("edges") @NonNull List<LinkEdge> edges) {
  public record LinkNode(@JsonProperty("id") @NonNull String id, @JsonProperty("label") @NonNull String label, @JsonProperty("url") @NonNull String url) {}

  public record LinkEdge(@JsonProperty("source") @NonNull String source, @JsonProperty("target") @NonNull String target) {}
}

File: webapp/src/domain/models/GraphData.ts
export interface LinkNode {
  id: string;
  label: string;
  url: string;
  // properties required by react-force-graph-2d
  x?: number;
  y?: number;
  val?: number; // for node size
  color?: string;
}

export interface LinkEdge {
  source: string | LinkNode; // react-force-graph-2d replaces string id with node object
  target: string | LinkNode;
}

export interface GraphData {
  nodes: LinkNode[];
  edges: LinkEdge[];
}
